### PR TITLE
[release-1.2] Memory Hotplug fixes and stabilization

### PR DESCRIPTION
**Pull request by acardace**
### Important before reviewing
The `setupMemoryHotplug` function lives in `pkg/virt-api/webhooks/defaults.go` while in v1.2 it lives in virt-controller, this makes the backport a bit harder to review, so just keep this in mind. The same goes for `setGuestMemory`, in the v1.2 we need 2 of them, one in the VMI validating webhook and one in virt-controller.

Original PR: https://github.com/kubevirt/kubevirt/pull/12128

### What this PR does
Before this PR:
Memory Hotplug causes a lot of issues and instability when `LiveUpdate` is used as `vmRolloutStrategy`

After this PR:
Memory hotplug does not cause issues anymore as it only gets setup when used (meaning when you trigger a memory hotplug), also KubeVirt now detects when a memory hotplug fails and marks the live-update migration as succeeded and sets the RestartRequired condition so that the VM owner can take action to make the VM changes effective after a restart. 

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
The virtio-mem device now gets attached only when hotplug is performed, and updated after the first hotplug.

The `HotMemoryChange` now gets set to `False` when a failed hotplug is detected, this also causes the `RestartRequired` condition.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Memory Hotplug fixes and stabilization
```

### CODE REVIEW

**Review by jean-edouard — COMMENTED** [review]
/approve
/hold
Did you consider adding the commit(s) that moved `setupMemoryHotplug()` or is that irrelevant?

File: pkg/liveupdate/memory/memory_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+	"kubevirt.io/kubevirt/pkg/liveupdate/memory"
+	"kubevirt.io/kubevirt/pkg/pointer"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/converter/vcpu"
+)
+
+var _ = Describe("LiveUpdate Memory", func() {
+	Context("Memory", func() {
+		var maxGuest resource.Quantity
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/liveupdate/memory/memory_test.go:37]: You're right, thank you.

File: pkg/liveupdate/memory/memory_test.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+					},
+				},
+				Status: v1.VirtualMachineInstanceStatus{
+					Memory: &v1.MemoryStatus{
+						GuestCurrent:   &currentGuestMemory,
+						GuestRequested: &currentGuestMemory,
+						GuestAtBoot:    &currentGuestMemory,
+					},
+				},
+			}
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/liveupdate/memory/memory_test.go:136]: Will do :+1:

**Comment by acardace** [comment]
> /approve /hold Did you consider adding the commit(s) that moved `setupMemoryHotplug()` or is that irrelevant?

I did, but the backport is really large already and the PR that moved that logic is also large, I didn't want to create unexpected mismatches in terms of code assumptions (like what should have been set and at which point) or behaviour, so I opted for what I consider a safer backport.

**Review by fossedihelm — APPROVED** [review]

**Review by fossedihelm — APPROVED** [review]

**Pull request merged**